Prometheus 2.19 retention not working

**What did you do?**
Our Prometheus is running in a pod on a Kubernetes cluster

The current setting of Prometheus is

$ ps auxwww | grep prometheus
    1 nobody   40:40 /bin/prometheus --config.file=/etc/prometheus/prometheus.yml --storage.tsdb.path=/prometheus --storage.tsdb.retention.size=40GB --log.level=info

Each time I just either extend with more disk space or recreate the brand new volume

**What did you expect to see?**
the WAL and chunk head space should stop growing when the disk space reaches the retention limit, like 40GB.

**What did you see instead? Under which circumstances?**
I see the tsdb keeps growing in size. now it's 99% again on a 60GB of FS.
And each I restart Prometheus, sometimes the disk space usage jumps by a few GB suddenly after it completing scanning WAL files. Not sure it recovers from some corruption or something.
Also the WAL scanning takes quite long; for example, I see it scanned 394 WAL segments in the most-recent of Prometheus restart. Each segment scanning took about 2-3 seconds, which cost 10+ mins to start Prometheus and get tsdb ready.

/prometheus $ df -h /prometheus
Filesystem                Size      Used Available Use% Mounted on
10.33.39.211:vol_6fc192fa0b17853d8fdac6eecf8ab53c
                         60.0G     59.6G    349.5M  99% /prometheus
/prometheus $ du -hs *
2.1M    01EGKWFED4J1GKWZ6173B0R9QG.tmp
4.0K    01EKMVV6503HMZKH1V8A40BV92.tmp
20.4M   01EKMZ00BMPHS4JG5S06NB5Y4D
9.2G    chunks_head
0       lock
4.0K    queries.active
48.2G   wal

**Environment**
Production

* System information:

$ uname -srm
Linux 3.10.0-1127.19.1.el7.x86_64 x86_64

* Prometheus version:

$ prometheus --version
prometheus, version 2.19.1 (branch: HEAD, revision: eba3fdcbf0d378b66600281903e3aab515732b39)
  build user:       root@62700b3d0ef9
  build date:       20200618-16:35:26
  go version:       go1.14.4

* Alertmanager version:

$ alertmanager --version
alertmanager, version 0.20.0 (branch: HEAD, revision: f74be0400a6243d10bb53812d6fa408ad71ff32d)
  build user:       root@00c3106655f8
  build date:       20191211-14:13:14
  go version:       go1.13.5

* Prometheus configuration file:
/etc/prometheus $ cat prometheus.yml
global:
  scrape_interval:     15s
  evaluation_interval: 15s

  external_labels:
    monitor: 'promswarm'

rule_files:
  - "swarm_node.rules.yml"
  - "swarm_task.rules.yml"
  - "demo_node.rules.yml"

alerting:
  alertmanagers:
  - static_configs:
    - targets:
      - alertmanager:9093

scrape_configs:
  - job_name: 'prometheus'
    static_configs:
      - targets: ['localhost:9090']


  - job_name: 'node-exporter'
    dns_sd_configs:
    - names:
      - 'node-exporter'
      type: 'A'
      port: 9100

  - job_name: 'blackbox'
    metrics_path: /probe
    params:
      module: [http_2xx]  # Look for a HTTP 200 response.
    static_configs:
      - targets:
        - http://grafana:3000 # Target to probe with Grafana
        - http://prometheus:9090 # Target to probe with Grafana
    relabel_configs:
      - source_labels: [__address__]
        target_label: __param_target
      - source_labels: [__param_target]
        target_label: instance
      - target_label: __address__
        replacement: blackbox:9115  # The blackbox exporter's real hostname:port


* Alertmanager configuration file:
It's kind of irrelevant to this issue because alertmanager is running in a separte pod in Kubernetes and the problem happens when alertmanager stops.

* Logs:


level=info ts=2020-10-02T14:15:10.621Z caller=main.go:337 msg="Starting Prometheus" version="(version=2.19.1, branch=HEAD, revision=eba3fdcbf0d378b66600281903e3aab515732b39)"
level=info ts=2020-10-02T14:15:10.621Z caller=main.go:338 build_context="(go=go1.14.4, user=root@62700b3d0ef9, date=20200618-16:35:26)"
level=info ts=2020-10-02T14:15:10.621Z caller=main.go:339 host_details="(Linux 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 11 19:12:04 EDT 2020 x86_64 prometheus-6475dc8c7b-js2kt (none))"
level=info ts=2020-10-02T14:15:10.621Z caller=main.go:340 fd_limits="(soft=1048576, hard=1048576)"
level=info ts=2020-10-02T14:15:10.621Z caller=main.go:341 vm_limits="(soft=unlimited, hard=unlimited)"
level=info ts=2020-10-02T14:15:10.642Z caller=main.go:678 msg="Starting TSDB ..."
level=info ts=2020-10-02T14:15:10.642Z caller=web.go:524 component=web msg="Start listening for connections" address=0.0.0.0:9090
level=info ts=2020-10-02T14:15:11.353Z caller=head.go:645 component=tsdb msg="Replaying WAL and on-disk memory mappable chunks if any, this may take a while"
level=info ts=2020-10-02T14:17:12.939Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=0 maxSegment=374
level=info ts=2020-10-02T14:17:15.065Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=1 maxSegment=374
level=info ts=2020-10-02T14:17:17.432Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=2 maxSegment=374
level=info ts=2020-10-02T14:17:19.632Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=3 maxSegment=374
level=info ts=2020-10-02T14:17:22.136Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=4 maxSegment=374
level=info ts=2020-10-02T14:17:24.352Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=5 maxSegment=374
level=info ts=2020-10-02T14:17:26.532Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=6 maxSegment=374
...
...
level=info ts=2020-10-02T14:31:06.330Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=369 maxSegment=374
level=info ts=2020-10-02T14:31:08.983Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=370 maxSegment=374
level=info ts=2020-10-02T14:31:11.307Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=371 maxSegment=374
level=info ts=2020-10-02T14:31:13.885Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=372 maxSegment=374
level=info ts=2020-10-02T14:31:15.387Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=373 maxSegment=374
level=info ts=2020-10-02T14:31:15.415Z caller=head.go:706 component=tsdb msg="WAL segment loaded" segment=374 maxSegment=374
level=info ts=2020-10-02T14:31:15.419Z caller=head.go:709 component=tsdb msg="WAL replay completed" duration=16m4.061215372s
level=info ts=2020-10-02T14:31:17.053Z caller=main.go:694 fs_type=65735546
level=info ts=2020-10-02T14:31:17.053Z caller=main.go:695 msg="TSDB started"
level=info ts=2020-10-02T14:31:17.053Z caller=main.go:799 msg="Loading configuration file" filename=/etc/prometheus/prometheus.yml
level=info ts=2020-10-02T14:31:17.068Z caller=main.go:827 msg="Completed loading of configuration file" filename=/etc/prometheus/prometheus.yml
level=info ts=2020-10-02T14:31:17.068Z caller=main.go:646 msg="Server is ready to receive web requests."
level=info ts=2020-10-02T14:31:26.401Z caller=compact.go:495 component=tsdb msg="write block" mint=1598384434664 maxt=1598385600000 ulid=01EKMVK1SW65N0ADF3HBZM6QJA duration=2.18090327s
level=info ts=2020-10-02T14:31:29.815Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=153.227589ms
level=error ts=2020-10-02T14:33:34.699Z caller=db.go:675 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: flush records: write /prometheus/wal/checkpoint.00000248.tmp/00000030: no space left on device"
level=info ts=2020-10-02T14:33:37.991Z caller=compact.go:495 component=tsdb msg="write block" mint=1598385600000 maxt=1598392800000 ulid=01EKMVQ26MDQ71SKEW4TC9TSV3 duration=2.291033222s
level=info ts=2020-10-02T14:33:46.420Z caller=compact.go:495 component=tsdb msg="write block" mint=1598385600000 maxt=1598392800000 ulid=01EKMVQAVQ0YZF182CXW3PEKQA duration=1.853869665s
level=info ts=2020-10-02T14:33:48.984Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=104.36129ms
...
...
level=info ts=2020-10-02T14:57:35.843Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=95.921353ms
level=error ts=2020-10-02T14:59:32.755Z caller=db.go:675 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: flush records: write /prometheus/wal/checkpoint.00000255.tmp/00000029: no space left on device"
level=info ts=2020-10-02T15:00:35.479Z caller=compact.go:495 component=tsdb msg="write block" mint=1598457600000 maxt=1598464800000 ulid=01EKMX8DBK6BJ5F511YSGYJMCM duration=2.723784021s
level=info ts=2020-10-02T15:00:42.354Z caller=compact.go:495 component=tsdb msg="write block" mint=1598457600000 maxt=1598464800000 ulid=01EKMX8MBBGPHB020JEH76ECKE duration=2.439220412s
level=info ts=2020-10-02T15:00:44.820Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=89.237455ms
level=error ts=2020-10-02T15:02:23.059Z caller=db.go:675 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: flush records: write /prometheus/wal/checkpoint.00000256.tmp/00000027: no space left on device"
level=info ts=2020-10-02T15:03:25.547Z caller=compact.go:495 component=tsdb msg="write block" mint=1598464800000 maxt=1598472000000 ulid=01EKMXDKNKJNZ0D52T7PFXCZ5W duration=2.488213148s
level=info ts=2020-10-02T15:03:31.453Z caller=compact.go:495 component=tsdb msg="write block" mint=1598464800000 maxt=1598472000000 ulid=01EKMXDT3H98F3WQVP3W4ZK1N0 duration=1.803627297s
level=info ts=2020-10-02T15:03:33.829Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=103.364827ms
level=error ts=2020-10-02T15:05:19.442Z caller=db.go:675 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: flush records: write /prometheus/wal/checkpoint.00000256.tmp/00000026: no space left on device"
level=info ts=2020-10-02T15:06:22.139Z caller=compact.go:495 component=tsdb msg="write block" mint=1598472000000 maxt=1598479200000 ulid=01EKMXJZXJHWY6CARXYQJ492B2 duration=2.696147591s
level=info ts=2020-10-02T15:06:29.350Z caller=compact.go:495 component=tsdb msg="write block" mint=1598472000000 maxt=1598479200000 ulid=01EKMXK7RE7RDD3X7HH69QTF8R duration=1.879378639s
level=info ts=2020-10-02T15:06:32.188Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=117.686065ms
level=error ts=2020-10-02T15:08:12.894Z caller=db.go:675 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: flush records: write /prometheus/wal/checkpoint.00000257.tmp/00000025: no space left on device"
level=info ts=2020-10-02T15:09:15.942Z caller=compact.go:495 component=tsdb msg="write block" mint=1598479200000 maxt=1598486400000 ulid=01EKMXR99Y3DP1QSE5HFSSP2B3 duration=3.047522436s
level=info ts=2020-10-02T15:09:24.526Z caller=compact.go:495 component=tsdb msg="write block" mint=1598479200000 maxt=1598486400000 ulid=01EKMXRJEAABBM4SD16BVV3TBE duration=2.275857153s
level=info ts=2020-10-02T15:09:27.105Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=97.779093ms
level=error ts=2020-10-02T15:11:02.704Z caller=db.go:675 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: flush records: write /prometheus/wal/checkpoint.00000258.tmp/00000026: no space left on device"
level=info ts=2020-10-02T15:12:05.150Z caller=compact.go:495 component=tsdb msg="write block" mint=1598486400000 maxt=1598493600000 ulid=01EKMXXF4GQHHA3WCJF2NY4FXH duration=2.44524409s
level=info ts=2020-10-02T15:12:11.292Z caller=compact.go:495 component=tsdb msg="write block" mint=1598486400000 maxt=1598493600000 ulid=01EKMXXNHP7N6JJYD8S3BPHEBC duration=2.021990675s
level=info ts=2020-10-02T15:12:13.941Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=87.655014ms
level=error ts=2020-10-02T15:13:47.998Z caller=db.go:675 component=tsdb msg="compaction failed" err="reload blocks: head truncate failed: create checkpoint: flush records: write /prometheus/wal/checkpoint.00000258.tmp/00000024: no space left on device"
level=info ts=2020-10-02T15:14:49.866Z caller=compact.go:495 component=tsdb msg="write block" mint=1598493600000 maxt=1598500800000 ulid=01EKMY2GHZQMX3VXBHPY27VYW0 duration=1.867483987s
level=info ts=2020-10-02T15:14:56.183Z caller=compact.go:495 component=tsdb msg="write block" mint=1598493600000 maxt=1598500800000 ulid=01EKMY2PGR1K8SHDWXKMGVQYTE duration=2.079231733s
level=info ts=2020-10-02T15:14:58.419Z caller=head.go:792 component=tsdb msg="Head GC completed" duration=92.539269ms

I'm also running into this issue.  I'm currently running v2.20.1. In my logs I noticed: 
```
panic: write header: write /prometheus/chunks_head/004616.tmp: no space left on device
```
I have retention size set to 325GB and my volume is 500Gi

--- a/tsdb/chunks/head_chunks.go
+++ b/tsdb/chunks/head_chunks.go
@@ -105,10 +105,6 @@ type ChunkDiskMapper struct {
 	// from which chunks are served till they are flushed and are ready for m-mapping.
 	chunkBuffer *chunkBuffer
 
-	// The total size of bytes in the closed files.
-	// Needed to calculate the total size of all segments on disk.
-	size atomic.Int64
-
 	// If 'true', it indicated that the maxt of all the on-disk files were set
 	// after iterating through all the chunks in those files.
 	fileMaxtSet bool
@@ -181,8 +177,6 @@ func (cdm *ChunkDiskMapper) openMMapFiles() (returnErr error) {
 		chkFileIndices = append(chkFileIndices, seq)
 	}
 
-	cdm.size.Store(int64(0))
-
 	// Check for gaps in the files.
 	sort.Ints(chkFileIndices)
 	if len(chkFileIndices) == 0 {
@@ -209,8 +203,6 @@ func (cdm *ChunkDiskMapper) openMMapFiles() (returnErr error) {
 		if v := int(b.byteSlice.Range(MagicChunksSize, MagicChunksSize+ChunksFormatVersionSize)[0]); v != chunksFormatV1 {
 			return errors.Errorf("%s: invalid chunk format version %d", files[i], v)
 		}
-
-		cdm.size.Add(int64(b.byteSlice.Len()))
 	}
 
 	return nil
@@ -371,7 +363,6 @@ func (cdm *ChunkDiskMapper) cut() (returnErr error) {
 		}
 	}()
 
-	cdm.size.Add(cdm.curFileSize())
 	cdm.curFileNumBytes.Store(int64(n))
 
 	if cdm.curFile != nil {
@@ -727,7 +718,6 @@ func (cdm *ChunkDiskMapper) deleteFiles(removedFiles []int) error {
 			cdm.readPathMtx.Unlock()
 			return err
 		}
-		cdm.size.Sub(int64(cdm.mmappedChunkFiles[seq].byteSlice.Len()))
 		delete(cdm.mmappedChunkFiles, seq)
 		delete(cdm.closers, seq)
 	}
@@ -766,8 +756,8 @@ func (cdm *ChunkDiskMapper) DeleteCorrupted(originalErr error) error {
 }
 
 // Size returns the size of the chunk files.
-func (cdm *ChunkDiskMapper) Size() int64 {
-	return cdm.size.Load() + cdm.curFileSize()
+func (cdm *ChunkDiskMapper) Size() (int64, error) {
+	return fileutil.DirSize(cdm.dir.Name())
 }
 
 func (cdm *ChunkDiskMapper) curFileSize() int64 {

--- a/tsdb/db_test.go
+++ b/tsdb/db_test.go
@@ -1237,7 +1237,7 @@ func TestSizeRetention(t *testing.T) {
 	// Add some data to the WAL.
 	headApp := db.Head().Appender(context.Background())
 	for _, m := range headBlocks {
-		series := genSeries(100, 10, m.MinTime, m.MaxTime)
+		series := genSeries(100, 10, m.MinTime, m.MaxTime+1)
 		for _, s := range series {
 			it := s.Iterator()
 			for it.Next() {
@@ -1256,8 +1256,12 @@ func TestSizeRetention(t *testing.T) {
 	blockSize := int64(prom_testutil.ToFloat64(db.metrics.blocksBytes)) // Use the actual internal metrics.
 	walSize, err := db.Head().wal.Size()
 	require.NoError(t, err)
-	// Expected size should take into account block size + WAL size
-	expSize := blockSize + walSize
+	cdmSize, err := db.Head().chunkDiskMapper.Size()
+	require.NoError(t, err)
+	require.NotZero(t, cdmSize)
+	// Expected size should take into account block size + WAL size + Head
+	// chunks size
+	expSize := blockSize + walSize + cdmSize
 	actSize, err := fileutil.DirSize(db.Dir())
 	require.NoError(t, err)
 	require.Equal(t, expSize, actSize, "registered size doesn't match actual disk size")
@@ -1270,7 +1274,20 @@ func TestSizeRetention(t *testing.T) {
 	blockSize = int64(prom_testutil.ToFloat64(db.metrics.blocksBytes)) // Use the actual internal metrics.
 	walSize, err = db.Head().wal.Size()
 	require.NoError(t, err)
-	expSize = blockSize + walSize
+	cdmSize, err = db.Head().chunkDiskMapper.Size()
+	require.NoError(t, err)
+	require.NotZero(t, cdmSize)
+	expSize = blockSize + walSize + cdmSize
+	actSize, err = fileutil.DirSize(db.Dir())
+	require.NoError(t, err)
+	require.Equal(t, expSize, actSize, "registered size doesn't match actual disk size")
+
+	// Truncate Chunk Disk Mapper and compare sizes.
+	require.NoError(t, db.Head().chunkDiskMapper.Truncate(900))
+	cdmSize, err = db.Head().chunkDiskMapper.Size()
+	require.NoError(t, err)
+	require.NotZero(t, cdmSize)
+	expSize = blockSize + walSize + cdmSize
 	actSize, err = fileutil.DirSize(db.Dir())
 	require.NoError(t, err)
 	require.Equal(t, expSize, actSize, "registered size doesn't match actual disk size")
@@ -1287,8 +1304,11 @@ func TestSizeRetention(t *testing.T) {
 	blockSize = int64(prom_testutil.ToFloat64(db.metrics.blocksBytes))
 	walSize, err = db.Head().wal.Size()
 	require.NoError(t, err)
+	cdmSize, err = db.Head().chunkDiskMapper.Size()
+	require.NoError(t, err)
+	require.NotZero(t, cdmSize)
 	// Expected size should take into account block size + WAL size
-	expSize = blockSize + walSize
+	expSize = blockSize + walSize + cdmSize
 	actRetentionCount := int(prom_testutil.ToFloat64(db.metrics.sizeRetentionCount))
 	actSize, err = fileutil.DirSize(db.Dir())
 	require.NoError(t, err)

--- a/tsdb/head.go
+++ b/tsdb/head.go
@@ -2373,7 +2373,8 @@ func (h *Head) Size() int64 {
 	if h.wal != nil {
 		walSize, _ = h.wal.Size()
 	}
-	return walSize + h.chunkDiskMapper.Size()
+	cdmSize, _ := h.chunkDiskMapper.Size()
+	return walSize + cdmSize
 }
 
 func (h *RangeHead) Size() int64 {

